SQL: forward warning headers to JDBC driver

Luegg opened the pull request:
Follow up on https://github.com/elastic/elasticsearch/pull/83943 to ensure that users of the JDBC drivers also can act on the warnings.

The change ensures that warnings in the response headers from the `_sql` API are available in `JdbcResultSet.getWarnings()`.

Commit:
forward warning headers to JDBC driver

elasticsearchmachine commented:
Hi @Luegg, I've created a changelog YAML for you.

Luegg commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/HttpClient.java:
```diff
@@ -82,10 +84,10 @@ public SqlQueryResponse basicQuery(String query, int fetchSize, boolean fieldMul
     }
 
-    public SqlQueryResponse query(SqlQueryRequest sqlRequest) throws SQLException {
+    public Tuple<SqlQueryResponse, List<String>> query(SqlQueryRequest sqlRequest) throws SQLException {
```
I'm not too fond of working with tuples everywhere I have to pass the warnings but this was the least intrusive way to get the warnings from the response headers. Since it's an internal client it probably doesn't matter too much but suggestions for solving this differently are welcome.

Commit:
Update docs/changelog/84499.yaml

Commit:
make headers lookup case-insensitive

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

astefan reviewed (approved):
LGTM. Left two comments.

Commit:
Merge origin/main into enhance/warningsInJdbc

Commit:
address review comments

bpintea commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/JreHttpUrlConnection.java:
```diff
@@ -183,6 +186,23 @@ public boolean head() throws ClientException {
+            // The linear scan below replicates the linear case-insensitive lookup of `URLConnection.getHeaderField(String)`.
+            List<String> values = new LinkedList<>();
+            int i = 0;
+            String value = con.getHeaderField(i);
+            while (value != null) {
```
Just noting that the function would stop iterating on an empty header value? This seems to be legal, though very corner case.

Luegg replied:
I hope that's never the case but the API does not forbid `null` headers. Changing the implementation to use `URLConnection.getHeaderFields()` 👍

bpintea reviewed (approved):
LGTM

costin reviewed (changes requested):
Looks good!
The pairing `ResultSet#clearWarnings` needs to be implemented as well - hence the request for changes.
An alternative to Tuple in 8.x could be a dedicated response class (maybe even a record in 8.x) to avoid the .v1() call or the large Tuple<SqlQueryResponse, List<String>> signature to `public ConnectionResponse ...` or something along those lines.

Commit:
Merge origin/main into enhance/warningsInJdbc

Commit:
fix `ResultSet.clearWarnings`

Commit:
replace Tuple with wrapper class

Commit:
safer method for getting headers from URLConnection

Luegg commented:
@costin good catch regarding `ResultSet#clearWarnings`. I've added the implementation and also replaced the tuple with a generic wrapper class (records are not available because we're stuck with Java 8 here).

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/Cursor.java:
```diff
@@ -28,4 +28,8 @@ default int columnSize() {
+
+    List<String> warnings();
+
+    Cursor withoutWarnings();
```
I'd prefer to remove the warnings instead of creating a new cursor instance.

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/DefaultCursor.java:
```diff
@@ -67,4 +77,13 @@ public void close() throws SQLException {
+
+    @Override
+    public Cursor withoutWarnings() {
+        return new DefaultCursor(client, cursor, columnInfos, rows, meta, Collections.emptyList());
```
`clearWarnings` method is easy to map the actual functionality and can simply clear the list with a check to be emptyList-friendly:
```
if (list.isEmpty() == false) { 
   list.clear()
}
```

Luegg replied:
`List#clear()` is not implemented in the list instance used here (it's unmodifiable).

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcResultSet.java:
```diff
@@ -54,7 +54,7 @@
     private final Calendar defaultCalendar;
 
     private final JdbcStatement statement;
-    private final Cursor cursor;
```
By using clearWarnings the Cursor is preserved as final.

costin commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcResultSet.java:
```diff
@@ -585,12 +585,21 @@ public InputStream getBinaryStream(String columnLabel) throws SQLException {
     @Override
     public void clearWarnings() throws SQLException {
         checkOpen();
+        cursor = cursor.withoutWarnings();
```
`cursor.clearWarnings()`

costin commented on x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/JreHttpUrlConnection.java:
```diff
@@ -183,6 +187,17 @@ public boolean head() throws ClientException {
 
+    private Function<String, List<String>> getHeaderFields(URLConnection con) {
+        return header -> {
+            for (Map.Entry<String, List<String>> entry : con.getHeaderFields().entrySet()) {
```
This method returns the first match - does headerFields normalizes keys? For example `Content-Type` and `ConteNT-TYPE` have their values under the same key or separate ones?

costin reviewed (approved):
Left some comments around styling, LGTM.

Commit:
make getHeaderFields more robust against inconsistent casing

Commit:
mutate cursor
